### PR TITLE
Initial WS support

### PR DESCRIPTION
**Pull request by ludeeus**
## Proposed change

Add websocket client, as MVP when core.state changes this will be used to signal the frontend about the changed data.

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (which adds functionality to the supervisor)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue:
- Link to documentation pull request:

## Checklist

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast supervisor tests`)
- [x] Tests have been added to verify that the new code works.

If API endpoints of add-on configuration are added/changed:

- [ ] Documentation added/updated for [developers.home-assistant.io][docs-repository]

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[docs-repository]: https://github.com/home-assistant/developers.home-assistant

### CODE REVIEW

File: supervisor/core.py [hunk 1]
```diff
@@ -55,6 +56,13 @@ def state(self, new_state: CoreState) -> None:
             )
         finally:
             self._state = new_state
+            if self.sys_core.state not in CLOSING_STATES:
+                self.sys_loop.call_soon_threadsafe(
+                    self.sys_loop.create_task,
+                    self.sys_homeassistant.websocket.async_supervisor_update_event(
+                        "info", {"state": new_state}
+                    ),
+                )
```
Comment (pvizeli) [inline supervisor/core.py:65]: Let's create a small helper on `self.sys_homeassistant.WebSocket.fire_websocket` which also handles errors with nice logging. Also, this setter is called from the event loop

File: supervisor/homeassistant/websocket.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    async def _get_ws_client(self) -> WSClient:
+        """Return a websocket client."""
+        await self.sys_homeassistant.api.ensure_access_token()
+        client = await WSClient.connect_with_auth(
+            self.sys_websession_ssl,
+            f"{self.sys_homeassistant.api_url}/api/websocket",
+            self.sys_homeassistant.api.access_token,
+        )
+        if client.ha_version < MIN_VERSION:
+            _LOGGER.warning("No WebSocket support untill core-%s", MIN_VERSION)
```
Comment (pvizeli) [inline supervisor/homeassistant/websocket.py:99]:
```suggestion
            _LOGGER.info("No WebSocket support untill core-%s", MIN_VERSION)
```

File: supervisor/core.py [hunk 3]
```diff
@@ -55,6 +56,13 @@ def state(self, new_state: CoreState) -> None:
             )
         finally:
             self._state = new_state
+            if self.sys_core.state not in CLOSING_STATES:
+                self.sys_loop.call_soon_threadsafe(
```
Comment (pvizeli) [inline supervisor/core.py]: is called from eventloop, `call_soon()`

File: supervisor/homeassistant/websocket.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+        try:
+            return await self._client.async_send_command(message)
+        except HomeAssistantAPIError as err:
+            raise HomeAssistantWSError from err
+
+    async def async_supervisor_update_event(
+        self, key: str, data: Optional[Dict[str, Any]] = None
+    ):
+        """Send a supervisor/event command."""
```
Comment (pvizeli) [inline supervisor/homeassistant/websocket.py:128]: Check the API state before we open a connection `sys_homeassistant.api.check_api_state()`

File: supervisor/homeassistant/websocket.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+        """Send a command with the WS client."""
+        if self.sys_core.state in CLOSING_STATES:
+            raise HomeAssistantWSNotSupported(
+                f"Can't execute in a ${self.sys_core.state} state"
+            )
+        if not self._client:
+            self._client = await self._get_ws_client()
+
+        if self._client.ha_version < MIN_VERSION:
+            raise HomeAssistantWSNotSupported
```
Comment (pvizeli) [inline supervisor/homeassistant/websocket.py:118]: Why we raise that there? We don't know if that command is supported. Only the Supervisor commands are not supported by this version

Comment (ludeeus) [reply supervisor/homeassistant/websocket.py:118]: True... I'll make a lookup table here

File: supervisor/homeassistant/websocket.py (hunk 4 above)
Comment (ludeeus) [reply supervisor/homeassistant/websocket.py:128]: That's done in `async_send_command`

**Review by pvizeli — APPROVED** [review]

**Pull request merged**